### ISSUE TITLE
Does not respect 'skip-tasklist' from wmctrl or gdevilspie

### ISSUE THREAD

**Issue by shiarus** [robint99/mate-dock-applet#140]
This bug secondly reported on Launchpad. I have also noticed another error that could benefit by a fix to this bug (GitHub: Generic icons which aren't an application appearing in dock #136). The bug is the title: Programs started that have rules created for them should be respected, whether those rules are user or system created. This could pose a security risk. If mate-dock-applet respected or searched against/validated its current running apps against the tasklist, it could narrow down or eliminate errors. To save resources this check would only have to be performed when the applet adds a new window.

I noticed this bug because I have some programs and scripts that start-up hidden from certain users (kids/students) to log and prevent user attempts to backdoor security measures (i.e. you cant type or copy to the clipboard the word 'proxy' or 'firewall' without the word(s) being replaced by the name of the user). For me, this applet is useful and I would like to make it available to other users, but this bug presents a security risk by not respecting those settings.

I have not elevated this to a security bug on launch pad, because I believe I have a special use case. However, if this bug could potentially lead to exposing/misidentify system programs that should remain hidden and revealing them to the user or granting a user elevated control when they shouldn't, then it poses an obvious and dangerous security risk. I welcome any admin to elevate this bug to a security risk if they believe my 'special use case' simply revealed a glaring security bug.

**Comment by shiarus** [comment]
It is possible to have the programs launch asynchronously with a script using devilspie.  If webkit (or some other library) is what you have to use to make this applet work, perhaps there is a command I can send to the service that provides that information to the applet to ignore or delist certain windows as a work around?

**Comment by robint99** [comment]
I'm working on  fix for this now. Should be up in a few days....

**Comment by robint99** [comment]
The applet is now respecting the 'skip-tasklist' flag so this should be fixed now :)

**Comment by shiarus** [comment]
Marking as closed, I also labeled as 'fix committed' to launch pad.  I may forget to ever update to 'fix released' unless I am shot an email from this bug report.

**Comment by shiarus** [comment]
I replaced 'src/dock.in' with your new version in my git clone, then when I compiled from source (--with-gtk3), and the applet was not functional (Couldn't add it to mate-panel).  Then I had an issue with the source code!  (Not from having is not from your changes, as I compiled your master without any modifications and had the same results.)  Here are the results from '$ python3 /usr/lib/mate-applets/mate-dock-applet/dock_applet.py':
Traceback (most recent call last):
  File "/usr/lib/mate-applets/mate-dock-applet/dock_applet.py", line 54, in <module>
    from Xlib.display import Display
ImportError: No module named 'Xlib'

I installed 'python3-xlib' as I already had 'python-xlib' and it solved this error (Mark this as a dependency).  However, did not solve my problem.  Again I tested (Didn't work) and ran '$ python3 /usr/lib/mate-applets/mate-dock-applet/dock_applet.py':
Traceback (most recent call last):
  File "/usr/lib/mate-applets/mate-dock-applet/dock_applet.py", line 67, in <module>
    import docked_app
  File "/usr/lib/mate-applets/mate-dock-applet/docked_app.py", line 60, in <module>
    gi.require_version("Bamf", "3")
  File "/usr/lib/python3/dist-packages/gi/__init__.py", line 102, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Bamf not available

This is beyond my experience, so I used some google-foo.  All I could do is apply the following fix after installing the applet with your modified 'src/dock.in'.  Something is obviously wrong with the install when compiling with gtk3.  I hope someone knows what anything else about that last traceback means, because I don't!

**Comment by shiarus** [comment]
OK, turns out I was missing a dependency: libglib2.0-dev.  When I added that, all was gold: '$ python3 /usr/lib/mate-applets/mate-dock-applet/dock_applet.py' returns nothing.  Unfortunately, mate-panel agrees that nothing is appropriate when I add it to the panel.  Now my debug tool is useless, help?

**Comment by robint99** [comment]
I've just compiled, installed and ran the latest version of the applet on a fully updated 16.04 virtual machine so I can confirm that it does work. 

If you're still getting the error about Bamf, install gir1.2-bamf-3 and that should fix it. Also, don't run .configure with the --wth-gtk3 flag. 16.04 is still on Gtk2 so you need only do ./configure --prefix=/usr

Try that and see how you get on. Any more problems just let me know :)

**Comment by shiarus** [comment]
Compiled from source, w/o gtk3.  It would seem mate-panel does not allow me to run this as Gtk2, because of something in PyGTK.  Because of this I am getting an import error and the applet will not add to mate-panel, no matter how it is configured.  What is different in your virtual box setting than my setup I do not know.

Output of '$ python2 /usr/lib/mate-applets/mate-dock-applet/dock_applet.py':

/usr/lib/python2.7/dist-packages/gi/overrides/Gtk.py:50: RuntimeWarning: You have imported the Gtk 2.0 module.  Because Gtk 2.0 was not designed for use with introspection some of the interfaces and API will fail.  As such this is not supported by the pygobject development team and we encourage you to port your app to Gtk 3 or greater. PyGTK is the recomended python module to use with Gtk 2.0
  warnings.warn(warn_msg, RuntimeWarning)
Traceback (most recent call last):
  File "/usr/lib/mate-applets/mate-dock-applet/dock_applet.py", line 57, in <module>
    from gi.repository import MatePanelApplet
  File "/usr/lib/python2.7/dist-packages/gi/importer.py", line 140, in load_module
    raise ImportError(e)
ImportError: Requiring namespace 'Gtk' version '3.0', but '2.0' is already loaded

**Comment by robint99** [comment]
It looks like something from your previous compilation is still hanging around. After compiling you did do
`sudo make install` did n't you?

In case this helps,  here are the full list of command you need to run to compile and install the applet after doing a git clone,

`aclocal`
`automake --add-missing`
`autoreconf`
`./configure --prefix=/usr`
`make`
`sudo make install`

**Comment by shiarus** [comment]
I have followed your steps, and manually removed all mate-dock-applet references from my system after purging.  I have tried a slew of configuration options, to no avail.  Running $ python3 '/usr/lib/mate-applets/mate-dock-applet/dock_applet.py' gives some errors.

/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:50: RuntimeWarning: You have imported the Gtk 2.0 module.  Because Gtk 2.0 was not designed for use with introspection some of the interfaces and API will fail.  As such this is not supported by the pygobject development team and we encourage you to port your app to Gtk 3 or greater. PyGTK is the recomended python module to use with Gtk 2.0
  warnings.warn(warn_msg, RuntimeWarning)
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/gi/importer.py", line 138, in load_module
    introspection_module = get_introspection_module(namespace)
  File "/usr/lib/python3/dist-packages/gi/module.py", line 275, in get_introspection_module
    module = IntrospectionModule(namespace, version)
  File "/usr/lib/python3/dist-packages/gi/module.py", line 122, in __init__
    repository.require(namespace, version)
gi.RepositoryError: Requiring namespace 'Gtk' version '3.0', but '2.0' is already loaded
